crates/parcel_core/src/plugin/transformer_plugin.rs

```diff
@@ -2,45 +2,59 @@ use parcel_filesystem::FileSystemRef;
 use std::fmt::Debug;
 use std::path::{Path, PathBuf};
 use std::rc::Rc;
+use std::sync::Arc;
 
-use crate::types::{Asset, SourceCode};
-use crate::types::{Dependency, SpecifierType};
+use crate::types::{Asset, Dependency, Environment, SourceCode, SpecifierType};
 
 pub struct ResolveOptions {
   /// A list of custom conditions to use when resolving package.json "exports" and "imports"
   pub package_conditions: Vec<String>,
-
   /// How the specifier should be interpreted
   pub specifier_type: SpecifierType,
 }
 
 /// A function that enables transformers to resolve a dependency specifier
 pub type Resolve = dyn Fn(PathBuf, String, ResolveOptions) -> Result<PathBuf, anyhow::Error>;
 
+/// A newly resolved file_path/code that needs to be transformed into an Asset
+pub struct InitialAsset {
+  pub file_path: PathBuf,
+  /// Dynamic code returned from the resolver for virtual files.
+  /// It is not set in most cases but should be respected when present.
+  pub code: Option<String>,
+  pub env: Arc<Environment>,
+  pub side_effects: bool,
+}
+
 /// The input to transform within the plugin
 ///
 /// Transformers may run against two distinguished scenarios:
-/// * File-paths that have just been discovered
+///
+/// * InitialAsset that have just been discovered
 /// * Outputs of previous transformation steps, which are in-place modified
 ///
 pub enum TransformationInput {
-  FilePath(PathBuf),
+  InitialAsset(InitialAsset),
   Asset(Asset),
 }
 
 impl TransformationInput {
   pub fn file_path(&self) -> &Path {
     match self {
-      TransformationInput::FilePath(path) => path,
+      TransformationInput::InitialAsset(raw_asset) => raw_asset.file_path.as_path(),
       TransformationInput::Asset(asset) => asset.file_path(),
     }
   }
 
-  pub fn read_source_code(&self, fs: FileSystemRef) -> anyhow::Result<Rc<SourceCode>> {
+  pub fn read_source_code(self, fs: FileSystemRef) -> anyhow::Result<Rc<SourceCode>> {
```

Comment:
why does it consume self? you'd want to store `Option<Rc<SourceCode>>` in the initial asset to prevent consuming things here.

Reply:
This was a typo actually I think 😅